### PR TITLE
improved type annotations in stats.py

### PR DESCRIPTION
**Pull request by silvandeleemput**
### The issue

I recently used the stats package of evalutils, but got some type annotation errors which were picked up by PyCharm.
Also running mypy on the package gave a lot of errors related to the stats.py package.
This PR aims to fix those.

### PR edits

* Extracted the `VOXELSPACING_TYPE`, since it can be quite different in appearance, all these formats are allowed:
  * `[1], 1, 2.0, [1,2,3], None, [1.0, 1.0], etc...`
* ~~Added type annotations to `distance_transform_edt_float32` method~~
* ~~pandas read_csv fix for pytests (3.7+), by explicitly setting the utf-8 encoding.~~
* ~~Refactored some code inside the `distance_transform_edt_float32` method to remove mypy errors.~~

### Additional notes
~~Pytests seem to work, but the pre-commit fails on the MyPy checks in the `distance_transform_edt_float32`~~

### CODE REVIEW

**Comment by jmsmkn** [comment]
Please fix the tests first

**Comment by silvandeleemput** [comment]
> Please fix the tests first

Yes, should be fixed now, I tested against the wrong version because my deploy settings were not properly configured.

**Comment by silvandeleemput** [comment]
Ok the remaining pytest errors are related to a version difference in `pandas`:
```
3.6, pandas==1.1.5
3.7, pandas==1.2.2
3.8, pandas==1.2.2

CSVLoader
 read_csv(fname, skipinitialspace=True).to_dict(
                orient="records"
            )


pandas.errors.ParserError: Error tokenizing data. C error: Expected 1 fields in line 7, saw 2
```

@james It should be a minor fix, I can create a new issue for it or attempt to solve it here.

**Comment by silvandeleemput** [comment]
@james only code coverage issues right now for a few exception cases (I did not introduce new ones) Do you want me to fix those, or can this be merged?

**Comment by silvandeleemput** [comment]
@jmsmkn in my opinion this PR should be ready to merge now:
* Added some edge case tests to improve coverage (tests pass)
* Fixed the pandas encoding issue that fixes the pytests in Python 3.7, 3.8 (tests pass)
* Replaced the `np.where(input, 1, 0).astype(np.int8)` in the `distance_transform_edt_float32` method with `input != 0` which is equivalent but is much faster and more memory lean. Some proof:

```
In [1]: import numpy as np

In [2]: data = np.random.random((1000,)) - 0.3

In [3]: data[:30] = 0

In [4]: data[30:60] = 1

In [5]: a = np.where(data, 1, 0)

In [6]: b = data != 0

In [7]: np.array_equal(a, b)

Out[7]: True

In [8]: %timeit data != 0
1.06 µs ± 18.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [9]: %timeit np.where(data, 1, 0)
3 µs ± 14.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [10]: %timeit data != 0
1.05 µs ± 2.79 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [11]: %timeit np.where(data, 1, 0)
2.99 µs ± 41 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

File: evalutils/stats.py [hunk 1]
```diff
@@ -8,13 +8,19 @@
...
-    return_indices=False,
-    distances=None,
-    indices=None,
+    input: ndarray,
+    sampling: VOXELSPACING_TYPE = None,
+    return_distances: bool = True,
+    return_indices: bool = False,
+    distances: Optional[ndarray] = None,
+    indices: Optional[ndarray] = None,
+    gc_byte_threshold: int = 100000000,
```
Comment (jmsmkn) [inline evalutils/stats.py:23]: Can you remove this extra argument? The function signature is intentionally the same as the one used in scipy.

File: evalutils/stats.py [hunk 2]
```diff
@@ -41,6 +47,9 @@ def distance_transform_edt_float32(  # noqa: C901
         Used for output of distance array, must be of type float64.
     indices : ndarray, optional
```
Comment (jmsmkn) [inline evalutils/stats.py:48]: If you're adding type hints to the function signature then remove them from the docstring as they are now out of sync.

File: evalutils/stats.py [hunk 3]
```diff
@@ -100,33 +109,41 @@ def distance_transform_edt_float32(  # noqa: C901
...
+            gc.collect()
 
-    garbage_collect = gc.collect if input.nbytes > 100e6 else lambda: None
     garbage_collect()
 
-    input = input.astype(np.int32)
+    input_data = input_data.astype(np.int32)
     garbage_collect()
 
+    sampling_arr: Optional[np.ndarray] = None
```
Comment (jmsmkn) [inline evalutils/stats.py:125]: You're changing the side effects of the function here, right? `ndarrays` are passed by reference right, so in the original function this transform will affect `sampling` in the outer scope. 

This function is here to do the _same_ thing as it does in scipy, but only with lower memory. I don't understand why you have made this changes with no issue, and lumped in with some typing changes.

**Review by jmsmkn — CHANGES_REQUESTED** [review]

File: evalutils/stats.py (hunk 1 above)
Comment (silvandeleemput) [reply evalutils/stats.py:23]: Hmmm yes, I can see how that's desirable. I introduced the argument only for testing the garbage collection conditions. 
If it is removed it cannot be tested in a reasonable time which in turn will reduce the coverage. I'll try to look for an alternative solution here.

File: evalutils/stats.py (hunk 2 above)
Comment (silvandeleemput) [reply evalutils/stats.py:48]: Good point, I'll do that.

File: evalutils/stats.py (hunk 3 above)
Comment (silvandeleemput) [reply evalutils/stats.py:125]: Ah, I was not aware that it should affect the sampling array in the outer scope as well.
I did these changes to remove MyPy errors. I'll have a look at fixing both.

File: evalutils/stats.py (hunk 1 above)
Comment (jmsmkn) [reply evalutils/stats.py:23]: You could declare this on the module level and then modify/patch that, but, I don't see testing those lines as that important.

File: evalutils/stats.py (hunk 3 above)
Comment (silvandeleemput) [reply evalutils/stats.py:125]: Ok, this is tricky, first of all, sampling is not passed in as a `ndarray` and it will not be changed in the outer scope as a side-effect. The `_ni_support._normalize_sequence` method will convert it to a list if it isn't already one and the np.asarray will convert it to a `ndarray`, hence the original reference is no longer relevant and otherwise no significant changes will be made.

However, I agree that this might be a bit too much for this PR, so as discussed I'll create a different one.

File: evalutils/stats.py (hunk 2 above)
Comment (silvandeleemput) [reply evalutils/stats.py]: Done.

File: evalutils/stats.py (hunk 1 above)
Comment (silvandeleemput) [reply evalutils/stats.py]: Ok, I reverted the method to its old form, so the `gc_byte_threshold` param is gone.

File: evalutils/stats.py (hunk 3 above)
Comment (silvandeleemput) [reply evalutils/stats.py]: Ok, I reverted the method body to its old form. MyPy errors are back, but this PR is purer in form.

**Comment by jmsmkn** [comment]
This PR cannot be merged with the pre-commit failures. I don't think that adding type hinting to this method is worth all this time, but I will leave it up to you.

**Review by jmsmkn — APPROVED** [review]
